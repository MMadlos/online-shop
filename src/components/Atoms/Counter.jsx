import { useState } from "react"

function Counter({ context, currentQuantity }) {
	// TODO - If counter is in ProductPage, minimum quantity is "1"
	// TODO - IF counter is in Cart, when counter is 0, it deletes the product from the Cart List

	// TODO - quantity es diferente en la pÃ¡gina de producto que en Cart.
	// -> En ProductPage() es del >= 1 independientemente de la cantidad que haya en el carrito
	// -> En CartPage() es = a la cantidad que haya en el carrito.

	const initialQuantity = context === "product" ? 1 : currentQuantity

	const [quantity, setQuantity] = useState(initialQuantity)

	function handleCounter(e) {
		e.preventDefault()
		e.stopPropagation()

		const buttonType = e.target.id

		setQuantity(buttonType === "increase" ? quantity + 1 : quantity - 1)
	}

	return (
		<div className="counter">
			<button
				id="decrease"
				onClick={handleCounter}
				disabled={quantity === 0 || (context === "product" && quantity === 1)}>
				-
			</button>
			<p>{quantity}</p>
			<button
				id="increase"
				onClick={handleCounter}>
				+
			</button>
		</div>
	)
}

export default Counter
