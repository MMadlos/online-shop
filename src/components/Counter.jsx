function Counter({ quantity }) {
	return (
		<div className="counter">
			<button id="increase"> - </button>
			<p>{quantity}</p>
			<button id="decrease"> + </button>
		</div>
	)
}

export default Counter
