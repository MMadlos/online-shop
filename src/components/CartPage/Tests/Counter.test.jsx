import { render, screen } from "@testing-library/react"
import { describe, it, expect, vi } from "vitest"

import Counter from "../Counter"
import userEvent from "@testing-library/user-event"

describe("Counter", () => {
	it("renders component", () => {
		const { container } = render(<Counter />)

		expect(container).toMatchSnapshot()
		// screen.debug()
	})

	it("renders quantity prop", () => {
		const quantity = 2

		render(<Counter quantity={quantity} />)
		expect(screen.getByText("2")).toBeInTheDocument()
		expect(screen.queryByText(/1/)).not.toBeInTheDocument()
		expect(screen.queryByText(/3/)).not.toBeInTheDocument()
	})
	it("renders buttons decrease and increase", () => {
		render(<Counter />)

		const buttons = screen.getAllByRole("button")

		expect(buttons.length).toBe(2)
		expect(buttons[0]).toBeInTheDocument()
		expect(buttons[1]).toBeInTheDocument()
		expect(buttons[0].id).toMatch("decrease")
		expect(buttons[1].id).toMatch("increase")
	})

	it("calls onClick function when button decrease is clicked", async () => {
		const onClickDecrease = vi.fn()
		const user = userEvent.setup()

		render(<Counter onClickButton={onClickDecrease} />)

		const decreaseButton = screen.getAllByRole("button")[0]

		await user.click(decreaseButton)
		expect(onClickDecrease).toHaveBeenCalled()
	})

	it("calls onClick function when button increase is clicked", async () => {
		const onClickIncrease = vi.fn()
		const user = userEvent.setup()

		render(<Counter onClickButton={onClickIncrease} />)

		const increaseButton = screen.getAllByRole("button")[1]

		await user.click(increaseButton)
		expect(onClickIncrease).toHaveBeenCalled()
	})

	it("doesn't call onClick fn when buttons are not clicked", () => {
		const onClick = vi.fn()

		render(<Counter onClickButton={onClick} />)

		expect(onClick).not.toHaveBeenCalled()
	})

	it("disables button decrease if quantity is 1", () => {
		render(<Counter quantity={1} />)
		const buttons = screen.getAllByRole("button")

		expect(buttons[0].disabled).toBe(true)
	})

	it("doesn't call onClick fn when button decrease is disabled and is clicked", async () => {
		const onClick = vi.fn()
		const user = userEvent.setup()
		const quantity = 1

		render(
			<Counter
				onClickButton={onClick}
				quantity={quantity}
			/>
		)

		const decreaseButton = screen.getAllByRole("button")[0]
		await user.click(decreaseButton)

		expect(onClick).not.toHaveBeenCalled()
	})
})
