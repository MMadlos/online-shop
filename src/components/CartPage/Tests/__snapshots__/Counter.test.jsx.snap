// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Counter > renders component 1`] = `
<div>
  <div
    class="counter"
  >
    <button
      id="decrease"
    >
      <i
        class="fa-solid fa-minus"
      />
    </button>
    <div
      class="quantity-container"
    >
      <p
        class="quantity"
      />
    </div>
    <button
      id="increase"
    >
      <i
        class="fa-solid fa-plus"
      />
    </button>
  </div>
</div>
`;
